### ISSUE TITLE
Report missing policy or params in status of ValidatingAdmissionPolicyBindings

### ISSUE THREAD

**Issue by jpbetz** [kubernetes/kubernetes#117263]
If a ValidatingAdmissionPolicyBinding references a non-existent policy in the `spec.policyName` field the binding doesn't do anything.

We have informally referred to a binding in this state as "mis-configured", but there is no way for a user to know that a binding is in this state short of attempting API requests that should be denied and noticing that they are allowed (which is clearly not a good way to check).

Part of the problem is that without a the `matchConditions` of the policy, the policy binding doesn't even know what API requests it should match.

We should report this state into the status of binding resources.

Note also that `spec.paramRef` can refer to a non-existant parameter resource. The behavior for this case is different. The binding is activated but the `param` variable will be null during evaluation of CEL expressions.  Should we report non-existant parameter resources to the status of the resource as well?

**Comment by shubham-singh-748** [comment]
hey can you please specify these followings:-
1. what type of resources need to create, user, group or any kind of resource
2. Required attributes for the resources, like name, id, etc..
3. Creating resource using appropriate APIs or console system.

**Comment by jpbetz** [comment]
cc @cici37 for visibility

**Comment by yufan-su** [comment]
Here is a brief work plan for the fix.

We will add a new Status field in ValidatingAdmissionPolicyBinding, like the following

```diff
type ValidatingAdmissionPolicyBinding struct {
    metav1.TypeMeta `json:",inline"`
    // Standard object metadata; More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata.
    // +optional
    metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
    // Specification of the desired behavior of the ValidatingAdmissionPolicyBinding.
    Spec ValidatingAdmissionPolicyBindingSpec `json:"spec,omitempty" protobuf:"bytes,2,opt,name=spec"`
+   // The status of the ValidatingAdmissionPolicyBinding, including warnings are set when the binded policy don't exist.
+   // Populated by the system.
+   // Read-only.
+   // +optional
+   Status ValidatingAdmissionPolicyBindingStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
}

type ValidatingAdmissionPolicyBindingStatus struct {
    // ObservedGeneration reflects the generation of the binding that was last processed
    // +optional
    ObservedGeneration int64
    
    // Conditions describe the state of the binding
    // +optional
    // +listType=map
    // +listMapKey=type
    Conditions []metav1.Condition
}
```

To populate the Status field, 
 - we will update the server side code [pkg/registry/admissionregistration/validatingadmissionpolicybinding](https://github.com/kubernetes/kubernetes/tree/master/pkg/registry/admissionregistration/validatingadmissionpolicybinding) to set the status during Create/Update
 -  we also need a controller to watch for the changes on `ValidatingAdmissionPolicy` resource, there are two options
     - Add a new controller parallel to the ValidatingAdmissionPolicy controller to set the status
     - Update the existing validatingadmissionpolicy-status-controller that ValidatingAdmissionPolicy uses to watch for this

@jpbetz @aaron-prindle, can you help to take a look at this plan and let me know if there is a preferred approach for the controller change? thank you.

**Comment by aaron-prindle** [comment]
LGTM to the work plan stated, ~~I believe extending the current controller is likely the best way to start here~~ (see comment below, creating a new controller is preferred).  @jpbetz do you have an opinion on the best practice for k/k related to whether to extend the existing ValidatingAdmissionPolicy status controller to operate on the ValidatingAdmissionPolicyBinding status vs creating a new ValidatingAdmissionPolicyBinding status controller?

**Comment by jpbetz** [comment]
A new controller is probably best.

**Comment by cici37** [comment]
I thought it would need a kep since it's a API change on a stable featue/API. If SIG lead is ok with moving forward without a kep, that's awesome ^^

**Comment by jpbetz** [comment]
> I thought it would need a kep since it's a API change on a stable featue/API. If SIG lead is ok with moving forward without a kep, that's awesome ^^

It will need a feature gate and KEP :(

But the KEP should be short/trivial.

**Comment by cici37** [comment]
> It will need a feature gate and KEP :(
> 
> But the KEP should be short/trivial.

Agreed. Just wanna remind @yufan-su since the original plan might be in this release^^
